pydantic_ai_slim/pydantic_ai/agent.py

```diff
@@ -1058,3 +1111,181 @@ def _prepare_result_schema(
             )
         else:
             return self._result_schema  # pyright: ignore[reportReturnType]
+
+
+@dataclasses.dataclass
+class AgentRunner(Generic[AgentDepsT, ResultDataT]):
+    graph_runner: GraphRunner[
+        _agent_graph.GraphAgentState, _agent_graph.GraphAgentDeps[AgentDepsT, Any], MarkFinalResult[ResultDataT]
+    ]
+
+    def __await__(self) -> Generator[Any, Any, AgentRunResult[AgentDepsT, ResultDataT]]:
+        """Run the agent graph until it ends, and return self."""
+
+        async def _run():
+            graph_run_result = await self.graph_runner
+            return AgentRunResult(graph_run_result)
+
+        return _run().__await__()
+
+    def __enter__(self) -> AgentRun[AgentDepsT, ResultDataT]:
+        graph_run = self.graph_runner.__enter__()
+        return AgentRun(graph_run)
+
+    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
+        self.graph_runner.__exit__(exc_type, exc_val, exc_tb)
+
+
+@dataclasses.dataclass
+class AgentRun(Generic[AgentDepsT, ResultDataT]):
+    """A stateful, iterable run of an agent."""
+
+    _graph_run: GraphRun[
+        _agent_graph.GraphAgentState, _agent_graph.GraphAgentDeps[AgentDepsT, Any], MarkFinalResult[ResultDataT]
+    ]
+
+    @property
+    def final_result(self) -> AgentRunResult[AgentDepsT, ResultDataT]:
+        graph_run_result = self._graph_run.final_result
+        return AgentRunResult(graph_run_result)
+
+    def __aiter__(
+        self,
+    ) -> AsyncIterator[
+        BaseNode[
+            _agent_graph.GraphAgentState,
+            _agent_graph.GraphAgentDeps[AgentDepsT, Any],
+            MarkFinalResult[ResultDataT],
+        ]
+        | End[MarkFinalResult[ResultDataT]]
+    ]:
+        return self
+
+    async def __anext__(
+        self,
+    ) -> (
+        BaseNode[
+            _agent_graph.GraphAgentState,
+            _agent_graph.GraphAgentDeps[AgentDepsT, Any],
+            MarkFinalResult[ResultDataT],
+        ]
+        | End[MarkFinalResult[ResultDataT]]
```

Comment:
Currently we yield `End`

```
<class 'pydantic_ai._agent_graph.ModelRequestNode'>
<class 'pydantic_ai._agent_graph.HandleResponseNode'>
<class 'pydantic_ai._agent_graph.FinalResultNode'>
<class 'pydantic_graph.nodes.End'>
```

I think we should stop at `FinalResultNode`.

Reply:
also it would be nice if all these node types had a `kind` attribute, so I didn't need extra imports and `isinstance` checks to detect the `FinalResultNode` etc.

Reply:
also it seems like right now `FinalResultNode` is effectively private, I think it needs to be public.

Reply:
I think we can probably get rid of `FinalResultNode`

Reply:
Looks like I already got rid of it, so it shouldn't show up any more.